File: uniconn-node/src/secure/message.ts
import type { IConn } from "../conn.js";
import { MSG_HELLO, MSG_HELLO_REPLY, MSG_DATA, MSG_ERROR, MAX_MESSAGE_SIZE } from "./constants.js";

// ── Marshalling helpers (big-endian) ─────────────────────

function writeU32BE(buf: Uint8Array, offset: number, val: number): void {
  buf[offset] = (val >>> 24) & 0xff;
  buf[offset + 1] = (val >>> 16) & 0xff;
  buf[offset + 2] = (val >>> 8) & 0xff;
  buf[offset + 3] = val & 0xff;
}

function readU32BE(buf: Uint8Array, offset: number): number {
  return (
    ((buf[offset] << 24) |
      (buf[offset + 1] << 16) |
      (buf[offset + 2] << 8) |
      buf[offset + 3]) >>>
    0
  );
}

function writeU16BE(buf: Uint8Array, offset: number, val: number): void {
  buf[offset] = (val >>> 8) & 0xff;
  buf[offset + 1] = val & 0xff;
}

function readU16BE(buf: Uint8Array, offset: number): number {
  return (buf[offset] << 8) | buf[offset + 1];
}

function writeU64BE(buf: Uint8Array, offset: number, val: bigint): void {
  const hi = Number((val >> 32n) & 0xffffffffn);
  const lo = Number(val & 0xffffffffn);
  writeU32BE(buf, offset, hi);
  writeU32BE(buf, offset + 4, lo);
}

function readU64BE(buf: Uint8Array, offset: number): bigint {
  const hi = BigInt(readU32BE(buf, offset));
  const lo = BigInt(readU32BE(buf, offset + 4));
  return (hi << 32n) | lo;
}

// ── Frame I/O ────────────────────────────────────────────

/** Read exactly `n` bytes from an IConn. */
async function readExact(conn: IConn, n: number): Promise<Uint8Array> {
  const result = new Uint8Array(n);
  let offset = 0;
  while (offset < n) {
    const chunk = result.subarray(offset);
    const read = await conn.read(chunk);
    if (read === 0) throw new Error("unexpected EOF");
    offset += read;
  }
  return result;
}

/** Read a framed message: [4B len][body]. */
export async function readFrame(conn: IConn): Promise<Uint8Array> {
  const lenBuf = await readExact(conn, 4);
  const len = readU32BE(lenBuf, 0);
  if (len > MAX_MESSAGE_SIZE) throw new Error(`frame too large: ${len}`);
  return readExact(conn, len);
}

/** Write a framed message: [4B len][body]. */
export async function writeFrame(conn: IConn, body: Uint8Array): Promise<void> {
  const frame = new Uint8Array(4 + body.length);
  writeU32BE(frame, 0, body.length);
  frame.set(body, 4);
  await conn.write(frame);
}

// ── Hello message ────────────────────────────────────────

export interface HelloMsg {
  type: number;
  publicKey: Uint8Array;
  payload: Uint8Array; // ek (HELLO) or ct (HELLO_REPLY)
  signature: Uint8Array;
}

export function marshalHello(msg: HelloMsg): Uint8Array {
  const bodyLen =
    1 + 2 + msg.publicKey.length + 2 + msg.payload.length + 2 + msg.signature.length;
  const buf = new Uint8Array(bodyLen);
  let off = 0;
  buf[off++] = msg.type;

  writeU16BE(buf, off, msg.publicKey.length);
  off += 2;
  buf.set(msg.publicKey, off);
  off += msg.publicKey.length;

  writeU16BE(buf, off, msg.payload.length);
  off += 2;
  buf.set(msg.payload, off);
  off += msg.payload.length;

  writeU16BE(buf, off, msg.signature.length);
  off += 2;
  buf.set(msg.signature, off);

  return buf;
}

export function unmarshalHello(body: Uint8Array): HelloMsg {
  if (body.length < 1) throw new Error("empty message body");
  let off = 0;
  const type = body[off++];

  if (off + 2 > body.length) throw new Error("truncated pubkey length");
  const pkLen = readU16BE(body, off);
  off += 2;
  if (off + pkLen > body.length) throw new Error("truncated pubkey data");
  const publicKey = body.slice(off, off + pkLen);
  off += pkLen;

  if (off + 2 > body.length) throw new Error("truncated payload length");
  const plLen = readU16BE(body, off);
  off += 2;
  if (off + plLen > body.length) throw new Error("truncated payload data");
  const payload = body.slice(off, off + plLen);
  off += plLen;

  if (off + 2 > body.length) throw new Error("truncated sig length");
  const sigLen = readU16BE(body, off);
  off += 2;
  if (off + sigLen > body.length) throw new Error("truncated sig data");
  const signature = body.slice(off, off + sigLen);

  return { type, publicKey, payload, signature };
}

// ── Data message ─────────────────────────────────────────

export function marshalData(
  timestamp: number,
  counter: bigint,
  ciphertext: Uint8Array,
): Uint8Array {
  const bodyLen = 1 + 4 + 8 + ciphertext.length;
  const buf = new Uint8Array(bodyLen);
  buf[0] = MSG_DATA;
  writeU32BE(buf, 1, timestamp);
  writeU64BE(buf, 5, counter);
  buf.set(ciphertext, 13);
  return buf;
}

export interface DataMsg {
  timestamp: number;
  counter: bigint;
  ciphertext: Uint8Array;
}

export function unmarshalData(body: Uint8Array): DataMsg {
  if (body.length < 13) throw new Error("DATA message too short");
  return {
    timestamp: readU32BE(body, 1),
    counter: readU64BE(body, 5),
    ciphertext: body.slice(13),
  };
}

// ── Error message ────────────────────────────────────────

export function marshalError(reason: string): Uint8Array {
  const reasonBytes = new TextEncoder().encode(reason);
  const buf = new Uint8Array(1 + 2 + reasonBytes.length);
  buf[0] = MSG_ERROR;
  writeU16BE(buf, 1, reasonBytes.length);
  buf.set(reasonBytes, 3);
  return buf;
}

export { writeU32BE, readU32BE, writeU64BE, readU64BE };

File: uniconn-py/uniconn/secure/message.py
"""Wire format marshal/unmarshal for USCP v1 messages (synchronous)."""

from __future__ import annotations

import struct
import socket

from .constants import MAX_MESSAGE_SIZE, MSG_DATA, MSG_ERROR, TIMESTAMP_SIZE, COUNTER_SIZE


def recv_exact(sock: socket.socket, n: int) -> bytes:
    """Receive exactly n bytes from a socket."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("unexpected EOF during recv_exact")
        buf.extend(chunk)
    return bytes(buf)


def read_frame_from_conn(conn) -> bytes:
    """Read a framed message from a Conn: [4B big-endian length][body]."""
    len_buf = bytearray(4)
    offset = 0
    while offset < 4:
        remaining = bytearray(4 - offset)
        n = conn.read(remaining)
        if n == 0:
            raise ConnectionError("unexpected EOF reading frame length")
        len_buf[offset:offset + n] = remaining[:n]
        offset += n
    length = struct.unpack("!I", len_buf)[0]
    if length > MAX_MESSAGE_SIZE:
        raise ValueError(f"frame too large: {length}")

    body = bytearray(length)
    offset = 0
    while offset < length:
        remaining = bytearray(length - offset)
        n = conn.read(remaining)
        if n == 0:
            raise ConnectionError("unexpected EOF reading frame body")
        body[offset:offset + n] = remaining[:n]
        offset += n
    return bytes(body)


def write_frame_to_conn(conn, body: bytes) -> None:
    """Write a framed message to a Conn: [4B big-endian length][body]."""
    frame = struct.pack("!I", len(body)) + body
    conn.write(frame)


def marshal_hello(
    msg_type: int, public_key: bytes, payload: bytes, signature: bytes
) -> bytes:
    """
    Serialize HELLO/HELLO_REPLY with framing:
    [4B frame_len][1B type][2B pk_len][pk][2B pl_len][pl][2B sig_len][sig]
    """
    body = bytearray()
    body.append(msg_type)
    body.extend(struct.pack("!H", len(public_key)))
    body.extend(public_key)
    body.extend(struct.pack("!H", len(payload)))
    body.extend(payload)
    body.extend(struct.pack("!H", len(signature)))
    body.extend(signature)

    frame = struct.pack("!I", len(body)) + body
    return bytes(frame)


def unmarshal_hello(body: bytes) -> dict:
    """Unmarshal HELLO or HELLO_REPLY from a frame body."""
    if not body:
        raise ValueError("empty message body")

    msg_type = body[0]
    off = 1

    pk_len = struct.unpack("!H", body[off : off + 2])[0]
    off += 2
    public_key = body[off : off + pk_len]
    off += pk_len

    pl_len = struct.unpack("!H", body[off : off + 2])[0]
    off += 2
    payload = body[off : off + pl_len]
    off += pl_len

    sig_len = struct.unpack("!H", body[off : off + 2])[0]
    off += 2
    signature = body[off : off + sig_len]

    return {
        "msg_type": msg_type,
        "public_key": bytes(public_key),
        "payload": bytes(payload),
        "signature": bytes(signature),
    }


def marshal_data(timestamp: int, counter: int, ciphertext: bytes) -> bytes:
    """Marshal DATA body: [1B type][4B ts][8B counter][ciphertext]."""
    body = bytearray()
    body.append(MSG_DATA)
    body.extend(struct.pack("!I", timestamp))
    body.extend(struct.pack("!Q", counter))
    body.extend(ciphertext)
    return bytes(body)


def marshal_error(reason: str) -> bytes:
    """Marshal ERROR body."""
    reason_bytes = reason.encode("utf-8")
    body = bytearray()
    body.append(MSG_ERROR)
    body.extend(struct.pack("!H", len(reason_bytes)))
    body.extend(reason_bytes)
    return bytes(body)
